src/3d/qgspointcloudlayer3drenderer.cpp

```diff
@@ -152,16 +153,26 @@ QgsPointCloudLayer3DRenderer *QgsPointCloudLayer3DRenderer::clone() const
 Qt3DCore::QEntity *QgsPointCloudLayer3DRenderer::createEntity( const Qgs3DMapSettings &map ) const
 {
   QgsPointCloudLayer *pcl = layer();
-  if ( !pcl || !pcl->dataProvider() || !pcl->dataProvider()->index() )
+  if ( !pcl || !pcl->dataProvider() || ( !pcl->dataProvider()->index() && pcl->dataProvider()->subIndexes().isEmpty() ) )
     return nullptr;
   if ( !mSymbol )
     return nullptr;
 
   const QgsCoordinateTransform coordinateTransform( pcl->crs(), map.crs(), map.transformContext() );
 
-  QgsPointCloudLayerChunkedEntity *entity = new QgsPointCloudLayerChunkedEntity( pcl->dataProvider()->index(), map, coordinateTransform, dynamic_cast<QgsPointCloud3DSymbol *>( mSymbol->clone() ), maximumScreenError(), showBoundingBoxes(),
-      static_cast< const QgsPointCloudLayerElevationProperties * >( pcl->elevationProperties() )->zScale(),
-      static_cast< const QgsPointCloudLayerElevationProperties * >( pcl->elevationProperties() )->zOffset(), mPointBudget );
+  Qt3DCore::QEntity *entity;
+  if ( pcl->dataProvider()->index() )
+  {
+    entity = new QgsPointCloudLayerChunkedEntity( pcl->dataProvider()->index(), map, coordinateTransform, dynamic_cast<QgsPointCloud3DSymbol *>( mSymbol->clone() ), maximumScreenError(), showBoundingBoxes(),
+        static_cast< const QgsPointCloudLayerElevationProperties * >( pcl->elevationProperties() )->zScale(),
+        static_cast< const QgsPointCloudLayerElevationProperties * >( pcl->elevationProperties() )->zOffset(), mPointBudget );
+  }
+  else
```

Comment:
shouldn't we check for `ContainSubIndexes` capability here before doing a virtual point cloud entity? What I am thinking is that if dataProvider()->index() is nullptr, it could be also that it is a "regular" point cloud file, but it just failed to get indexed?

Reply:
We're guarded by the early exit in line 156 for that. At this point we should be guaranteed that if index() is nullptr then there are subIndexes, but I agree that it can be made more explicit.